client-device-connection-infinispan/src/main/java/org/eclipse/hono/deviceconnection/infinispan/client/CacheBasedDeviceConnectionInfo.java

```diff
@@ -568,29 +590,60 @@ private String checkAdapterInstanceId(
         if (adapterInstanceId != null) {
             final AdapterInstanceStatus status = adapterInstanceStatusProvider.getStatus(adapterInstanceId);
             if (status == AdapterInstanceStatus.DEAD) {
-                LOG.debug("ignoring found adapter instance id, belongs to already terminated container [tenant: {}, device-id: {}, adapter-instance-id: {}]",
+                LOG.debug(
+                        "ignoring found adapter instance id, belongs to already terminated container [tenant: {}, device-id: {}, adapter-instance-id: {}]",
                         tenantId, deviceId, adapterInstanceId);
-                span.log("ignoring found adapter instance id [" + adapterInstanceId + "], belongs to already terminated container");
-                cache.remove(getAdapterInstanceEntryKey(tenantId, deviceId), adapterInstanceId)
+                span.log("ignoring found adapter instance id [" + adapterInstanceId
+                        + "], belongs to already terminated container");
+                final Future<Void> listenerResult;
+                if (deviceToAdapterMappingErrorListener != null) {
+                    listenerResult = deviceToAdapterMappingErrorListener.onObsoleteEntryFound(tenantId, deviceId,
```

Comment:
What is the motivation for having a `Future` returned here? Is it really needed/useful?
My intuition would be, that waiting for the `onObsoleteEntryFound` (i.e. `sendDisconnectEventIfNeeded`) result would unnecessarily delay deleting the cache entry. Thereby, for two commands arriving at the same time, there could potentially be 2 disconnected events being sent.

Reply:
I think it is matter of balance to use `Future` and block the call or not.
My intention was if the `onObsoleteEntryFound` call is not successful to not delete the cache entry.
This way on a subsequent command from north bound application there would be still event notification to the application.
Do you know how much time usually would take to deliver `ttd` event to the kafka?
Your point is getting higher priority only if the probability to happen is higher.
My understanding is that overlapping command calls could happen and there is no way to handle all of them properly.
Maybe the question is what is more probable:
- Subsequent overlapping command calls
- Error during the ttd event delivery

Reply:
An other point for using `Future` is the guarantee the message order of the `ttd` event and the later command response using `sendDeliveryFailureCommandResponseMessage`.
Although the event and the command response are using different topics and command response is not always used I think the order should be correct.

Reply:
Sending a disconnected event twice in the not so often happening case that there are 2 commands coming in at the same time for such a device, is probably not really a problem.
And yes, ensuring the order of `ttd` and `sendDeliveryFailureCommandResponse` messages would be a (small) benefit, possibly preventing the downstream application from doing an (unnecessary) command sending retry.
So, I'm ok with keeping the `Future` handling.

> My intention was if the `onObsoleteEntryFound` call is not successful to not delete the cache entry.
> This way on a subsequent command from north bound application there would be still event notification to the application.

But skipping the cache entry removal is not in the current code, right? I think we could do that here. It's quite an edge case though, I think. I could imagine it happening when sending the event fails because the Command Router is shutting down.

Reply:
Hmm, you are right the cache entry removal was not skipped, although that was my intention. It is corrected.